### PR TITLE
Adding more context to tensor.cpp method calls

### PR DESCRIPTION
**Pull request by alankao64**
Adding try-catch blocks to calls for methods in tensor.cpp for more context when a method fails.
Addresses Issue #216

### CODE REVIEW

File: src/contraction_utils.cpp [hunk 1]
```diff
@@ -102,14 +102,22 @@ ContractionData ContractionData::Initialize(
   long unsigned int max_size = (long unsigned int)pow(bond_dim, max_rank);
 
   // General-purpose scratch space (primarily used for tensor reordering).
-  data.scratch_.push_back(Tensor({""}, {max_size}));
+  try {
+    data.scratch_.push_back(Tensor({""}, {max_size}));
+  } catch (std::string err_msg) {
+    throw ERROR_MSG("Failed to call Tensor(). Error: ", err_msg);
```
Comment (alankao64) [inline src/contraction_utils.cpp]: should there be a new line before the error? `Tensor(). \nError: `

Comment (95-martin-orion) [reply src/contraction_utils.cpp]: I'd recommend trying a couple different things (line breaks, indentation) and going with whatever reads most clearly. One note of caution: we need to make sure that a chain of errors looks different from two separate errors. For example, we don't want to end up with this:
```
Failed to call <methodA>. Error:
Failed to call <methodB>. Error:
<original failure>
```
since it's hard to tell from a single line that it's part of another error.

File: src/contraction_utils.cpp [hunk 2]
```diff
@@ -171,7 +187,11 @@ void ContractionData::ContractGrid(
         Tensor& prev = get_scratch(op.expand.id);
         std::string result_id = result_space(patch_rank_[op.expand.id]);
         Tensor& result = get_scratch(result_id);
-        multiply(prev, next, result, get_scratch(kGeneralSpace).data());
+        try {
+          multiply(prev, next, result, get_scratch(kGeneralSpace).data());
```
Comment (95-martin-orion) [inline src/contraction_utils.cpp]: For `multiply`, `project`, and `reorder`, it would help to include the input tensor(s) in the error message.
(This can be a separate PR if you think that would be cleaner.)

Comment (alankao64) [reply src/contraction_utils.cpp]: I've got another branch where i'm improving error messages, i'll add these fixes to there.

**Comment by alankao64** [comment]
Sample error log:
```
ERROR (some_file.cpp: some_function:123) --> Failed to call another_function() Error:
      [ERROR (another_file.cpp: another_function:234) --> Error message thrown by another_function()]
```
